### ISSUE TITLE
Autorepeating test sets and other niceties from testing research?

### ISSUE THREAD

**Issue by robertfeldt** [IainNZ/BaseTestNext.jl#10]
Dear Ian and the team working on updates to Julia's support for testing. I just found out about these new features but must say things look great. Me and some of my research colleagues have been working on a new package for better test automation in Julia and was about to update that for Julia 0.4 before releasing it into the wild. However, when seeing the many changes you are working on it is tempting to rather try to fit in with what is coming in the future. :) So I hope you'd be willing to consider some ideas, proposals and maybe even code from us at some point.

I need to look more closely into the new base/test.jl implementation in 0.5 but it seems to me that a few, fairly small changes might make for a very powerful testing framework. The main thing I have in mind is to add a shouldrepeat method (or some better/shorter name) to the AbstractTestSet interface which would allow it to decide if tests should run another time before continuing (to the next testset or to the parent). Possibly also a start method which is called before starting to run the tests, but maybe a special version of the record method could be used also for this. I guess even the indication to run again could be given to the testset code if it returned something of a certain type if you really do not want to add more methods.

The idea behind this is that it would allow for adaptive decision on whether more testing is needed. In auto-testing with generated test data this would be very handy as one can use a sequential statistical test (!) to decide if enough evidence has been amassed to test a property of some stochastic/non-deterministic code etc. When combining this with our (award-winning! :) ) test data generation framework:

http://www.robertfeldt.net/publications/feldt_2013_godeltest.html

this would allow for very powerful forms of testing (property-based testing, Quickcheck style, for example).

An alternative way to accomplish this might be if the custom test set could be somehow referenced in the loop args in a testset loop. I guess it should be possible using the existing get_testset() but from looking at the code it seems the test set is generated inside the loop body so could not be used in the loop args, right? Also the code might not look very clean even if this could be achieved.

Sorry for this very long and long-winding issue post but wanted to get some indication of the general interest in getting Pull Requests to make the future Base.Test more powerful before we start thinking about how to merge/move over some of our testing code. We also have some code for producing more detailed failure messages to help pinpoint where the root cause might lie when executing a test as well as specific test macros that take decisions based on multiple runs of a testset but this might be taking things too far. But please give some indication of the willingness to extend the basic test interfacing and code; if we can find a few minimal changes that might increase power many times I think it would be worth it. Thanks for any response feedback you might have.

**Comment by robertfeldt** [comment]
Here is some example code showing some minimal changes to allow looping control from inside the custom test set. This was a quick'n'dirty to give a flavor of usefulness, in practice it should have much more refined logic to decide if to continue testing:

https://github.com/robertfeldt/BaseTestNext.jl/commit/243ea5f78e698ec685ee04a3ee4d39c4f831c917

**Comment by robertfeldt** [comment]
The example is now a bit better since it shows an example of finding a "rare", seeded bug with property-based testing and a (random) test data generator:

https://github.com/robertfeldt/BaseTestNext.jl/commit/971cb9a1c1ec0616bcfbd5137c5f604a1f9d40c9

**Comment by IainNZ** [comment]
Hi @robertfeldt,
This would be better at Julia's main repository. After this was merged into Base, this repo just serves as a "mirror" so people can use it on Julia 0.4.

**Comment by IainNZ** [comment]
I'd also say, 

>  So I hope you'd be willing to consider some ideas, proposals and maybe even code from us at some point.

I think people would be willing to consider code first, and ideas last. So if you start with some code showing some of your ideas (e.g. as a proof-of-concept PR), that'd be best.

**Comment by robertfeldt** [comment]
I did link to some concrete code and I selected this package since I thought it would be harder to get things into the main repo but ok, I'll press on and try to adapt some of our test tools to base/test.jl then. I guess my fear is that the code is likely to grow so maybe people would rather prefer it being its own package rather than go into base.

**Comment by IainNZ** [comment]
Maybe people would prefer it in a package, hard to say. Changing Base.Test to be more customizable though would probably go through fine. All I meant was that if you want people to see it, you should post it at JuliaLang/julia, only person really looking at this repo is me and I'm not doing any more work on testing for the foreseeable future.

**Comment by IainNZ** [comment]
and that this code just mirrors Base, so we aren't adding code here that isn't in Base.

**Comment by robertfeldt** [comment]
Ok, I did a RFC PR:

https://github.com/JuliaLang/julia/pull/14971/files

Comments appreciated.
